File: web/src/engineAnalysis.ts
// Chegga Web — per-move engine analysis (Phase 2)
//
// Ported from Chegga's own `app/services/engine_analysis.py::analyze_game`.
// For a game with n plies this runs n+1 engine analyses (one per board
// position, including the start and the final position), same as the
// backend — eval_before for the move played at ply i comes from position
// i's top line, eval_after comes from position i+1's top line (the REAL
// resulting position), not derived out of position i's MultiPV list.
//
// One real protocol difference from the Python port: raw UCI reports
// `score cp`/`score mate` from the side-to-move's own perspective at the
// analysed position, not White-relative the way python-chess's
// `PovScore.white()` does the conversion for us. This module does that
// conversion itself (`toWhiteRelativeCp`) before storing, so the stored
// fields keep the same White-relative convention as the backend's
// `MoveAnalysis` model.

import { Chess } from "chess.js";
import type { AnalysisLine, Engine } from "./engine";
import type { GameRecord, MoveAnalysisRecord } from "./db";
import { clocksByFen } from "./clockParser";
import { tagMove } from "./blunderTagger";
import { bandFor } from "./timePressure";

export const MATE_SCORE_CP = 100_000;
export const DISPLAY_CLAMP_CP = 1000; // caps mate-adjacent blowups so a single move can't dominate a chart

// Our own explicit, tunable convention — there is no industry-standard
// definition of these labels. Centipawn loss thresholds, mover's
// perspective, lowest to highest. Carried over unchanged from the
// backend so the two products agree on what "a blunder" means.
const CLASSIFICATION_THRESHOLDS: [number, string][] = [
  [10, "best"],
  [25, "excellent"],
  [50, "good"],
  [100, "inaccuracy"],
  [200, "mistake"],
];
const BLUNDER = "blunder";

export function classify(centipawnLoss: number): string {
  for (const [threshold, label] of CLASSIFICATION_THRESHOLDS) {
    if (centipawnLoss <= threshold) return label;
  }
  return BLUNDER;
}

export function gamePhase(ply: number, board: Chess): "opening" | "middlegame" | "endgame" {
  if (ply <= 20) return "opening";
  let nonPawnKingMaterial = 0;
  for (const row of board.board()) {
    for (const cell of row) {
      if (cell && cell.type !== "p" && cell.type !== "k") nonPawnKingMaterial += 1;
    }
  }
  return nonPawnKingMaterial <= 6 ? "endgame" : "middlegame";
}

/** cp equivalent for a mate score — mirrors python-chess's
 * `score.score(mate_score=MATE_SCORE_CP)`, which substitutes a large
 * magnitude for "mate in N" so cp math never has to special-case it (the
 * later clamp to DISPLAY_CLAMP_CP means the exact magnitude doesn't
 * matter, only the sign and that it dwarfs any real cp value). Exported
 * so any other cp/mate pair already in white-relative form (e.g.
 * analysisPanel.ts's live eval) can feed the same substitution without
 * going through an AnalysisLine. */
export function cpEquivalent(cp: number | undefined, mate: number | undefined): number {
  return mate !== undefined ? Math.sign(mate || 1) * MATE_SCORE_CP : (cp ?? 0);
}

function toWhiteRelativeCp(line: AnalysisLine | undefined, sideToMoveAtLine: "white" | "black"): number {
  if (!line) return 0;
  const raw = cpEquivalent(line.scoreCp, line.scoreMate);
  return sideToMoveAtLine === "white" ? raw : -raw;
}

function whiteRelativeMate(line: AnalysisLine | undefined, sideToMoveAtLine: "white" | "black"): number | undefined {
  if (!line || line.scoreMate === undefined) return undefined;
  return sideToMoveAtLine === "white" ? line.scoreMate : -line.scoreMate;
}

export interface AnalysisOptions {
  depth: number; // Chegga's backend default: 14
  multipv: number; // Chegga's backend default: 3
  movetimeMs?: number; // Chegga's backend default: 300 — same two-bound Limit as chess.engine.Limit(depth, time)
}

export const DEFAULT_ANALYSIS_OPTIONS: AnalysisOptions = { depth: 14, multipv: 3, movetimeMs: 300 };

export async function analyzeGame(
  engine: Engine,
  game: GameRecord,
  opts: AnalysisOptions = DEFAULT_ANALYSIS_OPTIONS,
  onProgress?: (done: number, total: number) => void,
): Promise<MoveAnalysisRecord[]> {
  const pgnGame = new Chess();
  pgnGame.loadPgn(game.pgn);
  const clockByFen = clocksByFen(pgnGame.getComments());
  const sanMoves = pgnGame.history();

  // Not every game starts from the standard position — a handicap/odds bot
  // game and a Chess.com "from position" game both carry their own start in
  // the PGN's FEN header (chess.js writes FEN + SetUp whenever a game didn't
  // begin from the normal setup). Replaying such a game from the standard
  // start throws on the first move that is only legal in the real position.
  const pgnStartFen = pgnGame.getHeaders().FEN;
  const board = pgnStartFen ? new Chess(pgnStartFen) : new Chess();
  const boardsByPly: Chess[] = [new Chess(board.fen())];
  const totalPositions = sanMoves.length + 1;
  const positionsInfo: AnalysisLine[][] = [await engine.analyse(board.fen(), opts)];
  onProgress?.(1, totalPositions);

  const uciMoves: string[] = [];
  for (const san of sanMoves) {
    const move = board.move(san);
    uciMoves.push(move.from + move.to + (move.promotion ?? ""));
    boardsByPly.push(new Chess(board.fen()));
    positionsInfo.push(await engine.analyse(board.fen(), opts));
    onProgress?.(positionsInfo.length, totalPositions);
  }

  const results: MoveAnalysisRecord[] = [];

  for (let i = 0; i < sanMoves.length; i++) {
    const boardBefore = boardsByPly[i];
    const sideToMove: "white" | "black" = boardBefore.turn() === "w" ? "white" : "black";
    const moverSign = sideToMove === "white" ? 1 : -1;

    const linesBefore = positionsInfo[i];
    const bestLine = linesBefore[0];
    const bestCpWhite = toWhiteRelativeCp(bestLine, sideToMove);
    const bestMateWhite = whiteRelativeMate(bestLine, sideToMove);

    const afterSideToMove: "white" | "black" = sideToMove === "white" ? "black" : "white";
    const afterLine = positionsInfo[i + 1][0];
    // A checkmated position has no legal moves, so the engine reports no
    // line for it at all. Reading that absence as "eval 0" made every
    // checkmating move look like it threw away a won game -- a delivered
    // mate graded as a 1000cp blunder, in the report, the annotated PGN and
    // every synced game's stats. The side to move in that position is the
    // side that got mated, so the mover won: score it as the mate it is.
    // A position with no line that isn't mate (stalemate, an insufficient-
    // material draw) really is 0 — a winning player who stalemates did lose
    // the win, and should be graded as having lost it.
    const afterCpWhite = afterLine
      ? toWhiteRelativeCp(afterLine, afterSideToMove)
      : boardsByPly[i + 1].isCheckmate()
        ? moverSign * MATE_SCORE_CP
        : 0;
    const afterMateWhite = afterLine ? whiteRelativeMate(afterLine, afterSideToMove) : undefined;

    const bestBeforeMover = bestCpWhite * moverSign;
    const actualAfterMover = afterCpWhite * moverSign;
    const centipawnLoss = Math.min(Math.max(0, bestBeforeMover - actualAfterMover), DISPLAY_CLAMP_CP);

    const uci = uciMoves[i];
    const moveRankLine = linesBefore.find((line) => line.pv[0] === uci);
    const moveRank = moveRankLine?.multipv;
    const bestPvMove = bestLine?.pv[0];

    const ply = i + 1;
    const san = sanMoves[i];
    let bestMoveSan: string | undefined;
    if (bestPvMove) {
      const probe = new Chess(boardBefore.fen());
      try {
        const m = probe.move({ from: bestPvMove.slice(0, 2), to: bestPvMove.slice(2, 4), promotion: bestPvMove.slice(4) || undefined });
        bestMoveSan = m.san;
      } catch {
        bestMoveSan = undefined; // engine's PV move didn't parse against this position — leave unset rather than guess
      }
    }

    const classification = classify(centipawnLoss);
    const clockSeconds = clockByFen.get(boardsByPly[i + 1].fen());

    results.push({
      gameId: game.chessComUuid,
      ply,
      sideToMove,
      fenBefore: boardBefore.fen(),
      san,
      uci,
      evalBeforeCp: bestCpWhite,
      evalBeforeMate: bestMateWhite,
      evalAfterCp: afterCpWhite,
      evalAfterMate: afterMateWhite,
      bestMoveUci: bestPvMove,
      bestMoveSan,
      centipawnLoss,
      moveRank,
      classification,
      gamePhase: gamePhase(ply, boardBefore),
      blunderTag: tagMove({
        fenBefore: boardBefore.fen(),
        uci,
        san,
        bestMoveSan,
        evalBeforeMate: bestMateWhite,
        evalAfterMate: afterMateWhite,
        sideToMove,
        classification,
      }),
      clockSeconds,
      timePressureBand: bandFor(clockSeconds, game.timeControl),
    });
  }

  return results;
}

File: archive/backend/app/services/engine_analysis.py
"""Walks stored games through Stockfish and records per-move analysis.

For a game with n plies this runs n+1 engine analyses (one per board
position, including the start and the final position) with MultiPV, not 2n.
eval_before for the move played at ply i comes from position i's top line;
eval_after comes from position i+1's top line -- the REAL resulting
position, not derived out of position i's MultiPV list, which would miss
the true eval whenever the played move falls outside the top-N candidates.

Analysis is currently supported for standard chess only. Chess960/variant
games are ingested in Phase 1 but skipped here -- a documented limitation,
not a guess at variant handling that was never actually exercised.
"""
import io
import logging
from datetime import datetime

import chess
import chess.engine
import chess.pgn
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import Settings
from app.models.game import Game
from app.models.move_analysis import MoveAnalysis
from app.services.blunder_tagger import tag_move
from app.services.clock_parser import parse_clocks_by_ply
from app.services.time_pressure_service import band_for

logger = logging.getLogger(__name__)

MATE_SCORE_CP = 100_000
DISPLAY_CLAMP_CP = 1000  # caps mate-adjacent blowups so a single move can't dominate a chart

SUPPORTED_RULES = {"chess"}

# Our own explicit, tunable convention -- there is no industry-standard
# definition of these labels. Centipawn loss thresholds, mover's
# perspective, lowest to highest.
_CLASSIFICATION_THRESHOLDS = [
    (10, "best"),
    (25, "excellent"),
    (50, "good"),
    (100, "inaccuracy"),
    (200, "mistake"),
]
_BLUNDER = "blunder"


def classify(centipawn_loss: int) -> str:
    for threshold, label in _CLASSIFICATION_THRESHOLDS:
        if centipawn_loss <= threshold:
            return label
    return _BLUNDER


def game_phase(ply: int, board: chess.Board) -> str:
    if ply <= 20:
        return "opening"
    non_pawn_king_material = sum(
        len(board.pieces(piece_type, color))
        for color in (chess.WHITE, chess.BLACK)
        for piece_type in (chess.KNIGHT, chess.BISHOP, chess.ROOK, chess.QUEEN)
    )
    return "endgame" if non_pawn_king_material <= 6 else "middlegame"


def _white_relative_cp(score: chess.engine.PovScore) -> tuple[int | None, int | None]:
    white_score = score.white()
    return white_score.score(mate_score=MATE_SCORE_CP), white_score.mate()


def _as_sorted_lines(info: chess.engine.InfoDict | list[chess.engine.InfoDict]) -> list[chess.engine.InfoDict]:
    lines = info if isinstance(info, list) else [info]
    return sorted(lines, key=lambda line: line.get("multipv", 1))


def analyze_game(engine: chess.engine.SimpleEngine, game_row: Game, settings: Settings) -> list[MoveAnalysis]:
    parsed = chess.pgn.read_game(io.StringIO(game_row.pgn))
    if parsed is None:
        raise ValueError(f"Could not parse stored PGN for game {game_row.id}")

    moves = list(parsed.mainline_moves())
    clocks_by_ply = parse_clocks_by_ply(parsed)
    limit = chess.engine.Limit(depth=settings.stockfish_depth, time=settings.stockfish_movetime_ms / 1000)
    multipv = settings.stockfish_multipv

    board = parsed.board()
    boards_by_ply = [board.copy()]
    positions_info = [_as_sorted_lines(engine.analyse(board, limit, multipv=multipv))]
    for move in moves:
        board.push(move)
        positions_info.append(_as_sorted_lines(engine.analyse(board, limit, multipv=multipv)))
        boards_by_ply.append(board.copy())

    results: list[MoveAnalysis] = []
    for i, move in enumerate(moves):
        board_before = boards_by_ply[i]
        mover_sign = 1 if board_before.turn == chess.WHITE else -1

        lines_before = positions_info[i]
        best_line = lines_before[0]
        best_cp_white, best_mate_white = _white_relative_cp(best_line["score"])

        after_cp_white, after_mate_white = _white_relative_cp(positions_info[i + 1][0]["score"])

        best_before_mover = (best_cp_white or 0) * mover_sign
        actual_after_mover = (after_cp_white or 0) * mover_sign
        centipawn_loss = min(max(0, best_before_mover - actual_after_mover), DISPLAY_CLAMP_CP)

        move_rank = next(
            (line.get("multipv", 1) for line in lines_before if line.get("pv") and line["pv"][0] == move),
            None,
        )
        best_move = best_line.get("pv", [None])[0]

        ply = i + 1
        side_to_move = "white" if board_before.turn == chess.WHITE else "black"
        san = board_before.san(move)
        best_move_san = board_before.san(best_move) if best_move else None
        classification = classify(centipawn_loss)
        clock_seconds = clocks_by_ply.get(ply)

        results.append(
            MoveAnalysis(
                game_id=game_row.id,
                ply=ply,
                side_to_move=side_to_move,
                fen_before=board_before.fen(),
                san=san,
                uci=move.uci(),
                eval_before_cp=best_cp_white,
                eval_before_mate=best_mate_white,
                eval_after_cp=after_cp_white,
                eval_after_mate=after_mate_white,
                best_move_uci=best_move.uci() if best_move else None,
                best_move_san=best_move_san,
                centipawn_loss=centipawn_loss,
                move_rank=move_rank,
                classification=classification,
                game_phase=game_phase(ply, board_before),
                blunder_tag=tag_move(
                    fen_before=board_before.fen(),
                    uci=move.uci(),
                    san=san,
                    best_move_san=best_move_san,
                    eval_before_mate=best_mate_white,
                    eval_after_mate=after_mate_white,
                    side_to_move=side_to_move,
                    classification=classification,
                ),
                clock_seconds=clock_seconds,
                time_pressure_band=band_for(clock_seconds, game_row.time_control),
            )
        )

    return results


def analyze_pending_games(db: Session, settings: Settings, limit: int | None = None) -> int:
    if not settings.stockfish_path:
        raise RuntimeError("STOCKFISH_PATH is not set -- fill it in in backend/.env first")

    stmt = (
        select(Game)
        .where(Game.analyzed.is_(False), Game.rules.in_(SUPPORTED_RULES))
        .order_by(Game.end_time.desc())  # newest first: useful data fast, older history backfills over time
    )
    if limit:
        stmt = stmt.limit(limit)
    pending = list(db.scalars(stmt))

    if not pending:
        logger.info("No pending games to analyze")
        return 0

    # One long-lived engine process reused across the whole batch --
    # Stockfish's startup cost makes per-game spawning wasteful.
    engine = chess.engine.SimpleEngine.popen_uci(settings.stockfish_path)
    try:
        engine.configure({"Threads": settings.stockfish_threads})
        analyzed_count = 0
        for game_row in pending:
            try:
                move_rows = analyze_game(engine, game_row, settings)
            except Exception:
                logger.exception("Failed to analyze game %s (%s)", game_row.id, game_row.chess_com_uuid)
                db.rollback()
                continue

            for row in move_rows:
                db.add(row)
            game_row.analyzed = True
            game_row.analyzed_at = datetime.utcnow()
            db.commit()  # transactional per game -- a crash mid-game must never leave it falsely marked done
            analyzed_count += 1
            logger.info("Analyzed game %s: %d moves", game_row.chess_com_uuid, len(move_rows))
        return analyzed_count
    finally:
        engine.quit()
